### ISSUE TITLE
Incorrect inner_join on columns with negative i8:s

### ISSUE THREAD

**Issue by RalfNorthman** [pola-rs/polars#3729]
#### What language are you using?

**Rust**

#### Which feature gates did you use?

lazy, dtype-i8

#### Have you tried latest version of polars?

- [yes]

#### What version of polars are you using?

0.22.8

#### What operating system are you using polars on?

Windows 10

#### What language version are you using

Rust stable 1.61.0

#### Describe your bug.

Joining two dataframes on columns with negative i8:s gives unexpected result.

#### What are the steps to reproduce the behavior?

```rust
use mimalloc::MiMalloc;
use polars::prelude::*;

#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;

fn main() -> Result<()> {
    let df1 = df![
        "a" => [-6_i8, -3, 0],
    ]?
    .lazy();

    let df2 = df![
        "b1" => [-6_i8, -4, -2, 0],
        "b2" => [-6_i8, -4, -2, 0],
    ]?
    .lazy();

    let df = df1.inner_join(df2, col("a"), col("b1")).collect()?;

    println!("{df}");

    Ok(())
}
```
#### What is the actual behavior?

```
shape: (7, 2)
┌─────┬─────┐
│ a   ┆ b2  │
│ --- ┆ --- │
│ i8  ┆ i8  │
╞═════╪═════╡
│ -6  ┆ -6  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ -3  ┆ -6  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ -6  ┆ -4  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ -3  ┆ -4  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ -6  ┆ -2  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ -3  ┆ -2  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ 0   ┆ 0   │
└─────┴─────┘
```

#### What is the expected behavior?

With i32:s instead of i8:s the result is as expected:
```
shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b2  │
│ --- ┆ --- │
│ i32 ┆ i32 │
╞═════╪═════╡
│ -6  ┆ -6  │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ 0   ┆ 0   │
└─────┴─────┘
```
Also columns with positive i8:s works as expected:
```
shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b2  │
│ --- ┆ --- │
│ i8  ┆ i8  │
╞═════╪═════╡
│ 6   ┆ 6   │
├╌╌╌╌╌┼╌╌╌╌╌┤
│ 0   ┆ 0   │
└─────┴─────┘
```

**Comment by ritchie46** [comment]
Thanks a lot for reporting this! I am surprised this has taken so long to be discovered. I have a fix in #3730.

I am not yet able to patch this to crates.io. I will have to wait until we do a new arrow2 release. Could you point to `master` for the time being?

**Comment by RalfNorthman** [comment]
That was quick! Thanks!

Yeah, it's fine pointing to master for now. Pointing to master also fixed my bigger problem, the one I was searching for a minimal example for when I found the bug in this issue. I made some joins and got a bunch of these:
```
thread '<unnamed>' panicked at 'attempt to add with overflow', C:\Users\ralfn\.cargo\registry\src\github.com-1ecc6299db9ec823\polars-core-0.22.7\src\frame\hash_join\mod.rs:145:12
```
But not on master! On master my project does not panic.

Thank you for this great crate!
